elastic_service/elastic_handler.go

```diff
@@ -0,0 +1,165 @@
+	}
+	return nil
+}
+
+func (e *ElasticHandler) AddBook(ctx context.Context, body []byte) (string, error) {
+	res, err := e.Client.Index().
+		Index(consts.Index).
+		BodyString(string(body)).
+		Do(ctx)
+	if err != nil {
+		return "", errors.Wrap(err, fmt.Sprintf("couldn't update book with body: %s", string(body)))
+	}
+	return res.Id, err
+}
+func (e *ElasticHandler) GetBook(ctx context.Context, id string) (src *models.Book, err error) {
+	get, err := e.Client.Get().
+		Index(consts.Index).
+		Id(id).
+		Do(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("couldn't get book form server with id: %s", id))
+	}
+
+	var book models.Book
+	if err = json.Unmarshal(get.Source, &book); err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("couldn't unmarshal result of book with id: %s, in getBoook function ", id))
+	}
+	return &book, err
+}
+
+func (e *ElasticHandler) DeleteBook(ctx context.Context, id string) error {
+	_, err := e.Client.Delete().
+		Index(consts.Index).
+		Id(id).
+		Do(ctx)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("couldn't delete book with id: %s", id))
+	}
+	return err
+}
+
+func buildSearchQuery(client *elastic.Client, title string, author string, priceRange string) (s *elastic.SearchService, err error) {
+
+	var from, to int
+	if priceRange != "" {
+		priceSplit := strings.Split(priceRange, "-")
+		if len(priceSplit) != 2 {
+			return nil, errors.New("failed to pars `price_range` field")
+		}
+		if from, err = strconv.Atoi(priceSplit[0]); err != nil {
+			return nil, errors.New("failed to pars `price_range` field")
+		}
+		if to, err = strconv.Atoi(priceSplit[1]); err != nil {
+			return nil, errors.New("failed to pars `price_range` field")
+		}
+	}
+
+	if from < consts.MinPrice {
+		return nil, errors.New(fmt.Sprintf("illegal price range, price should be higher than %d", consts.MinPrice))
+	}
+	if to > consts.MaxPrice {
+		return nil, errors.New(fmt.Sprintf("illegal price range, price should be lower than %d", consts.MaxPrice))
+	}
+
+	all := elastic.NewMatchAllQuery()
+	builder := client.Search().Index(consts.Index).Query(all).Pretty(true)
+	q := elastic.NewBoolQuery()
+
+	if title != "" {
+		q.Must(elastic.NewMatchQuery(consts.Title, title))
+	}
+	if author != "" {
+		q.Must(elastic.NewMatchQuery(consts.Author, author))
+	}
+	if to != 0 {
+		q.Must(elastic.NewRangeQuery(consts.Price).From(from).To(to))
+	}
+	builder = builder.Query(q)
+	return builder, nil
+}
+
+func (e *ElasticHandler) Search(title string, author string, priceRange string) (res []models.Book, err error) {
+
+	builder, err := buildSearchQuery(e.Client, title, author, priceRange)
+	if err != nil {
+		return nil, err
+	}
+	searchResult, err := builder.Pretty(true).Size(config.MaxQueryResults).Do(context.TODO())
+
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("search failed for title: %s, author: %s, priceRange: %s", title, author, priceRange))
+	}
+	res = []models.Book{}
```

Comment:
This declaration is redundant (you decide on the function signeture)

Reply:
you need to handle it not here, and at an upper level (if err == nill && res == nil .. ) or remove the `res` from the function signature